crypto/bp/bp_fp12.c

```diff
@@ -0,0 +1,881 @@
+        goto err;
+    if (!FP6_neg(group, r->f[1], a->f[1]))
+        goto err;
+    if (!FP6_mul(group, r->f[1], r->f[1], t0, ctx))
+        goto err;
+
+    ret = 1;
+
+ err:
+    FP6_free(t0);
+    FP6_free(t1);
+    return ret;
+}
+
+int FP12_conj(const BP_GROUP *group, FP12 *r, const FP12 *a)
+{
+    FP6_copy(r->f[0], a->f[0]);
+    if (!FP6_neg(group, r->f[1], a->f[1]))
+        return 0;
+    return 1;
+}
+
+int FP12_frb(const BP_GROUP *group, FP12 *r, const FP12 *a, BN_CTX *ctx)
+{
+    int ret = 0;
+
+    if (!FP2_conj(group, r->f[0]->f[0], a->f[0]->f[0]))
+        goto err;
+    if (!FP2_conj(group, r->f[1]->f[0], a->f[1]->f[0]))
+        goto err;
+    if (!FP2_conj(group, r->f[0]->f[1], a->f[0]->f[1]))
+        goto err;
+    if (!FP2_conj(group, r->f[1]->f[1], a->f[1]->f[1]))
+        goto err;
+    if (!FP2_conj(group, r->f[0]->f[2], a->f[0]->f[2]))
+        goto err;
+    if (!FP2_conj(group, r->f[1]->f[2], a->f[1]->f[2]))
+        goto err;
+    if (!FP2_mul_frb(group, r->f[1]->f[0], r->f[1]->f[0], 1, ctx))
+        goto err;
+    if (!FP2_mul_frb(group, r->f[0]->f[1], r->f[0]->f[1], 2, ctx))
+        goto err;
+    if (!FP2_mul_frb(group, r->f[1]->f[1], r->f[1]->f[1], 3, ctx))
+        goto err;
+    if (!FP2_mul_frb(group, r->f[0]->f[2], r->f[0]->f[2], 4, ctx))
+        goto err;
+    if (!FP2_mul_frb(group, r->f[1]->f[2], r->f[1]->f[2], 5, ctx))
+        goto err;
+    ret = 1;
+ err:
+    return ret;
+}
+
+int FP12_cyc(const BP_GROUP *group, FP12 *r, const FP12 *a, BN_CTX *ctx)
+{
+    FP12 *t = NULL;
+    int ret = 0;
+
+    if ((t = FP12_new()) == NULL)
+        goto err;
+
+    if (!FP12_inv(group, t, a, ctx))
+        goto err;
+    if (!FP12_conj(group, r, a))
+        goto err;
+    if (!FP12_mul(group, r, r, t, ctx))
+        goto err;
+
+    if (!FP12_frb(group, t, r, ctx))
+        goto err;
+    if (!FP12_frb(group, t, t, ctx))
+        goto err;
+    if (!FP12_mul(group, r, r, t, ctx))
+        goto err;
+
+    ret = 1;
+ err:
+    FP12_free(t);
+    return ret;
+}
+
+int FP12_exp_cyc(const BP_GROUP *group, FP12 *r, const FP12 *a,
+                 const BIGNUM *b, BN_CTX *ctx)
+{
+    int i, ret = 0;
+    FP12 *t = { NULL };
+
+    if ((t = FP12_new()) == NULL)
+        goto err;
+
+    if (!FP12_copy(t, a))
+        goto err;
+
+    for (i = BN_num_bits(b) - 2; i >= 0; i--) {
+        if (!FP12_sqr_cyc(group, t, t, ctx))
+            goto err;
+        if (BN_is_bit_set(b, i)) {
+            if (!FP12_mul(group, t, t, a, ctx))
+                goto err;
+        }
+    }
+
+    if (!FP12_copy(r, t))
+        goto err;
+
+    ret = 1;
+ err:
+    FP12_free(t);
+    return ret;
+}
+
+int FP12_exp_pck(const BP_GROUP *group, FP12 *r, const FP12 *a,
+                 const BIGNUM *b, BN_CTX *ctx)
+{
+    int i, j, w, ret = 0;
+
+    w = 0;
+    for (i = 1; i < BN_num_bits(b); i++) {
+        if (BN_is_bit_set(b, i)) {
+            w++;
+        }
+    }
+
+    FP12 *t[w];
```

Comment:
Need to initialize these, otherwise invalid free if one of the `FP12_new()` calls fails.